### PR TITLE
Add 5 different task source channels

### PR DESCRIPTION
**Pull request by KiChjang**
This is **not** complete. I really need feedback right away since I felt that the direction I'm heading is very wrong.

Partial #7959.

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/8871)

### CODE REVIEW

**Comment by KiChjang** [comment]
First of all, did I get the naming right? Is it supposed to be called a task queue or a task source?

I chose task queue because it made more sense to type `DOMManipulationTaskQueue.send(msg)` rather than `DOMManipulationTaskSource.send(msg)`.

**Comment by KiChjang** [comment]
Secondly, I'm currently simply using `MainThreadScriptChan`'s implementations and making 5 differently named copies of them. Is this correct? I'm thinking that I should instead write a trait called `GenericTaskSource` instead and have all 5 task sources to `impl` it, but if I do so, are these task sources/task queues still supposed to be a wrapper around `Sender`s?

**Comment by KiChjang** [comment]
I'm also running into problems while trying to eliminate `MainThreadScriptChan`.

First of all, some background information - `MainThreadScriptChan` is simply a wrapper around `Sender<MainThreadScriptMsg>`. `MainThreadScriptMsg` is an enum that contains an algebraic data type called `Common(CommonScriptMsg)`. When an instance of `MainThreadScriptChan` calls the `send` method, it takes in a `CommonScriptMsg` parameter and sends `MainThreadScriptMsg:::Common(CommonScriptMsg)` to the receiver.

Right now, in `window.rs`, `fn main_thread_script_chan` returns `&Sender<MainThreadScriptMsg>`. I'm guessing this is because callers of this function actually want to send a `MainThreadScriptMsg` instead of a `CommonScriptMsg` because there are other messages aside from `Common::(CommonScriptMsg)`s that you can send as a `MainThreadScriptMsg`.

Now here's the problem: do I still return `Box<ScriptChan + Send>` when I call `window.dom_manipulation_task_source()` or any other task source? Or do I return a `&Sender<MainThreadScriptMsg>` just like `MainThreadScriptChan`?

**Comment by KiChjang** [comment]
Ok, ignore what I said earlier in this thread; this is only a problem because I'm breaking up #7959 in manageable pieces. In a later PR, I plan to redesign the entire messaging system, so that `MainThreadScriptMsg` will die.

File: components/script/script_task.rs [hunk 1]
```diff
@@ -504,7 +609,7 @@ impl ScriptTaskFactory for ScriptTask {
             PipelineNamespace::install(state.pipeline_namespace_id);
             let roots = RootCollection::new();
             let _stack_roots_tls = StackRootTLS::new(&roots);
-            let chan = MainThreadScriptChan(script_chan);
+            let chan = MainThreadScriptChan(script_chan.clone());
```
Comment (jdm) [inline components/script/script_task.rs]: I think this isn't used now?

**Comment by jdm** [comment]
@bors-servo: delegate+
This looks like a useful start to me as written. We can merge this with my single comment addressed.

File: components/script/script_task.rs (hunk 1 above)
Comment (KiChjang) [inline components/script/script_task.rs]: It's used for `channel_for_reporter`, defined just below this line.

**Pull request merged**